From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 21 Dec 2022 16:07:06 +0900
Subject: [PATCH] Lithium-CachedHashCode


diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index 0c348b366623e350393f035d760adc6ee4142687..eb3d528889890fde4d8dcbe3399710c963a8a4cd 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -667,11 +667,18 @@ public class Block extends BlockBehaviour implements ItemLike {
         private final BlockState first;
         private final BlockState second;
         private final Direction direction;
+        private final int hash; // Fusion - Lithium: CachedHashCode
 
         public BlockStatePairKey(BlockState self, BlockState other, Direction facing) {
             this.first = self;
             this.second = other;
             this.direction = facing;
+            // Fusion start - Lithium: CachedHashCode - Initialize the object's hashcode and cache it
+            int hash = this.first.hashCode();
+            hash = 31 * hash + this.second.hashCode();
+            hash = 31 * hash + this.direction.hashCode();
+            this.hash = hash;
+            // Fusion end
         }
 
         public boolean equals(Object object) {
@@ -686,12 +693,14 @@ public class Block extends BlockBehaviour implements ItemLike {
             }
         }
 
+        // Fusion start - Lithium: CachedHashCod
+        /**
+         * @reason Uses the cached hashcode
+         * @author JellySquid
+         */
         public int hashCode() {
-            int i = this.first.hashCode();
-
-            i = 31 * i + this.second.hashCode();
-            i = 31 * i + this.direction.hashCode();
-            return i;
+            return this.hash;
         }
+        // Fusion end
     }
 }
