From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Fri, 28 Oct 2022 09:19:40 +0900
Subject: [PATCH] Lithium-EntityFastPowderSnowCheck

Mirai - etil2jz - GPL 3.0
Lithium - CaffeineMC  - GPL 3.0

diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index 55e8736e7136d4003b71badcf1b8b53dd22258c5..e593b6c9bee23da1e1040ea12097bae2c750f78d 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -205,6 +205,7 @@ public class FusionConfig {
     public static boolean lithiumCollectionsPrecomputeShapeArrays = true;
     public static boolean lithiumEntityFastElytraCheck = true;
     public static boolean lithiumEntityFastHandSwing = true;
+    public static boolean lithiumEntityFastPowderSnowCheck = true;
 
     private static void lithiumConfig() {
         setComment("lithium", "[ Lithium ] General-Purpose Optimization Mod");
@@ -218,5 +219,6 @@ public class FusionConfig {
         lithiumCollectionsPrecomputeShapeArrays = getBoolean("lithium.collections.precompute-shape-arrays", lithiumCollectionsPrecomputeShapeArrays) && lithiumEnable;
         lithiumEntityFastElytraCheck = getBoolean("lithium.entity.fast-elytra-check", lithiumEntityFastElytraCheck) && lithiumEnable;
         lithiumEntityFastHandSwing = getBoolean("lithium.entity.fast-hand-swing", lithiumEntityFastHandSwing) && lithiumEnable;
+        lithiumEntityFastPowderSnowCheck = getBoolean("lithium.entity.fast-powder-snow-check", lithiumEntityFastPowderSnowCheck) && lithiumEnable;
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index cd7d5554f0a38759e768351010e2df8e095b7ae4..b6f55802e105b60e1fc3162b07799980556a72bc 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -576,11 +576,11 @@ public abstract class LivingEntity extends Entity {
     }
 
     protected void tryAddFrost() {
-        if (!this.getBlockStateOnLegacy().isAir()) {
+        if (com.github.ipecter.fusion.FusionConfig.lithiumEntityFastPowderSnowCheck || !this.getBlockStateOnLegacy().isAir()) { // Fusion - Lithium: Entity Fast PowderSnow Check
             int i = this.getTicksFrozen();
 
             if (i > 0) {
-                AttributeInstance attributemodifiable = this.getAttribute(Attributes.MOVEMENT_SPEED);
+                AttributeInstance attributemodifiable = com.github.ipecter.fusion.FusionConfig.lithiumEntityFastPowderSnowCheck && this.getBlockStateOnLegacy().isAir() ? null : this.getAttribute(Attributes.MOVEMENT_SPEED); // Fusion - Lithium: Entity Fast PowderSnow Check
 
                 if (attributemodifiable == null) {
                     return;
