From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 14 Nov 2022 17:03:22 +0900
Subject: [PATCH] SportPaper-FixHeadRotatePacketSpam

Mirai - etil2jz - GPL 3.0
SportPaper - Electroid - GPL 3.0

diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index 6fb326f066b04abb6aa57a18f24dee0ebfc66e13..aa042a466af7ce45d04d06996937e0c323b6c932 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -171,9 +171,11 @@ public class FusionWorldConfig {
     }
 
     public boolean packetDontSendUselessEntityPackets = true;
+    public boolean packetFixEntityHeadRotatePacketSpam = true;
 
     private void packetSettings() {
         packetDontSendUselessEntityPackets = getBoolean("packet.dont-send-useless-entity-packets", packetDontSendUselessEntityPackets);
+        packetFixEntityHeadRotatePacketSpam = getBoolean("packet.fix-entity-head-rotate-packet-spam", packetFixEntityHeadRotatePacketSpam);
     }
 
     private void packetSettingsSetup() {
diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 88765672b90936718d39483da2e762a54eddf1a8..56c76388c878a50bd8a0cacc7b8f61efaf16d2aa 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -343,8 +343,12 @@ public class ServerEntity {
         }
 
         // CraftBukkit start - Fix for nonsensical head yaw
-        this.yHeadRotp = Mth.floor(this.entity.getYHeadRot() * 256.0F / 360.0F);
-        consumer.accept(new ClientboundRotateHeadPacket(this.entity, (byte) this.yHeadRotp));
+        // Fusion start - SportPaper: Fix Head Rotate Packet Spam
+        if (this.level.fusionConfig.packetFixEntityHeadRotatePacketSpam && this.entity instanceof LivingEntity) {
+            this.yHeadRotp = Mth.floor(this.entity.getYHeadRot() * 256.0F / 360.0F);
+            consumer.accept(new ClientboundRotateHeadPacket(this.entity, (byte) this.yHeadRotp));
+        }
+        // Fusion end
         // CraftBukkit end
 
         if (this.entity instanceof LivingEntity) {
