From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 14 Nov 2022 17:11:44 +0900
Subject: [PATCH] Slice-NoFirePacketWithFireResistance

Slice - Cryptite - MIT

diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index aa042a466af7ce45d04d06996937e0c323b6c932..4f941f262195d66244ee7b9d3c98c9bfc160fd08 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -172,10 +172,12 @@ public class FusionWorldConfig {
 
     public boolean packetDontSendUselessEntityPackets = true;
     public boolean packetFixEntityHeadRotatePacketSpam = true;
+    public boolean packetNoFirePacketWithFireResistance = true;
 
     private void packetSettings() {
         packetDontSendUselessEntityPackets = getBoolean("packet.dont-send-useless-entity-packets", packetDontSendUselessEntityPackets);
         packetFixEntityHeadRotatePacketSpam = getBoolean("packet.fix-entity-head-rotate-packet-spam", packetFixEntityHeadRotatePacketSpam);
+        packetNoFirePacketWithFireResistance = getBoolean("packet.no-fire-packet-with-fire-resistance", packetNoFirePacketWithFireResistance);
     }
 
     private void packetSettingsSetup() {
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 048509cc6fb824dad6c30c668b0c96eca3257d94..228222eb30bc8d1e25c3108a51daa3203dfa8d23 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -879,7 +879,13 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
 
         this.checkOutOfWorld();
         if (!this.level.isClientSide) {
-            this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            // Fusion start - Slice: No FirePacket With FireResistance
+            if (this instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0 && !(this.level.fusionConfig.packetNoFirePacketWithFireResistance && livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE)));
+            } else {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            }
+            // Fusion end
         }
 
         this.firstTick = false;
