From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 11 Jan 2023 10:25:16 +0900
Subject: [PATCH] C2ME-PooledFeatureContext


diff --git a/src/main/java/com/ishland/c2me/opts/allocs/common/PooledFeatureContext.java b/src/main/java/com/ishland/c2me/opts/allocs/common/PooledFeatureContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..ab3f0515b0fab016f73a0a46f80a10ddef0b91da
--- /dev/null
+++ b/src/main/java/com/ishland/c2me/opts/allocs/common/PooledFeatureContext.java
@@ -0,0 +1,69 @@
+package com.ishland.c2me.opts.allocs.common;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.level.WorldGenLevel;
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
+import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext;
+import net.minecraft.world.level.levelgen.feature.configurations.FeatureConfiguration;
+
+import java.util.Optional;
+
+public class PooledFeatureContext<FC extends FeatureConfiguration> extends FeaturePlaceContext<FC> {
+
+    public static final ThreadLocal<SimpleObjectPool<PooledFeatureContext<?>>> POOL = ThreadLocal.withInitial(() -> new SimpleObjectPool<>(unused -> new PooledFeatureContext<>(), unused -> {}, 2048));
+
+    private Optional<ConfiguredFeature<?, ?>> topFeature;
+    private WorldGenLevel level;
+    private ChunkGenerator chunkGenerator;
+    private RandomSource random;
+    private BlockPos origin;
+    private FC config;
+
+    public PooledFeatureContext() {
+        super(null, null, null, null, null, null);
+    }
+
+    public void reInit(Optional<ConfiguredFeature<?, ?>> feature, WorldGenLevel world, ChunkGenerator generator, RandomSource random, BlockPos origin, FC config) {
+        this.topFeature = feature;
+        this.level = world;
+        this.chunkGenerator = generator;
+        this.random = random;
+        this.origin = origin;
+        this.config = config;
+    }
+
+    public void reInit() {
+        this.topFeature = null;
+        this.level = null;
+        this.chunkGenerator = null;
+        this.random = null;
+        this.origin = null;
+        this.config = null;
+    }
+
+    public Optional<ConfiguredFeature<?, ?>> topFeature() {
+        return this.topFeature;
+    }
+
+    public WorldGenLevel level() {
+        return this.level;
+    }
+
+    public ChunkGenerator chunkGenerator() {
+        return this.chunkGenerator;
+    }
+
+    public RandomSource random() {
+        return this.random;
+    }
+
+    public BlockPos origin() {
+        return this.origin;
+    }
+
+    public FC config() {
+        return this.config;
+    }
+}
\ No newline at end of file
