From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 14 Nov 2022 15:57:40 +0900
Subject: [PATCH] C2ME-CachedString

Mirai - etil2jz - GPL 3.0
C2ME - RelativityMC - MIT

diff --git a/src/main/java/net/minecraft/resources/ResourceLocation.java b/src/main/java/net/minecraft/resources/ResourceLocation.java
index 7017dd42f832d928f1008a05f01701667d951644..746aead83ed071f7188659b561433200e1706611 100644
--- a/src/main/java/net/minecraft/resources/ResourceLocation.java
+++ b/src/main/java/net/minecraft/resources/ResourceLocation.java
@@ -27,6 +27,7 @@ public class ResourceLocation implements Comparable<ResourceLocation> {
     public static final String REALMS_NAMESPACE = "realms";
     protected final String namespace;
     protected final String path;
+    private String cachedString = null; // Fusion - C2ME: Cached String
 
     protected ResourceLocation(String[] id) {
         this.namespace = StringUtils.isEmpty(id[0]) ? "minecraft" : id[0];
@@ -99,7 +100,16 @@ public class ResourceLocation implements Comparable<ResourceLocation> {
 
     @Override
     public String toString() {
-        return this.namespace + ":" + this.path;
+        // Fusion start - C2ME: Cached String
+        /**
+         * @author ishland
+         * @reason cache toString
+         */
+        if (this.cachedString != null) return this.cachedString;
+        final String s = this.namespace + ":" + this.path;
+        this.cachedString = s;
+        return s;
+        // Fusion end
     }
 
     @Override
