From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 9 Jan 2023 12:19:42 +0900
Subject: [PATCH] Lithium-Util-EntitySectionPosition


diff --git a/src/main/java/net/minecraft/world/level/entity/EntitySection.java b/src/main/java/net/minecraft/world/level/entity/EntitySection.java
index 5dcb7f9cab097990148f5a7c4ccbe1556afdd514..cc07585885496b83cae274238dec2b0534ee46bd 100644
--- a/src/main/java/net/minecraft/world/level/entity/EntitySection.java
+++ b/src/main/java/net/minecraft/world/level/entity/EntitySection.java
@@ -13,12 +13,23 @@ public class EntitySection<T extends EntityAccess> {
     private static final Logger LOGGER = LogUtils.getLogger();
     private final ClassInstanceMultiMap<T> storage;
     private Visibility chunkStatus;
+    private long pos; // Fusion - Lithium: Util EntitySectionPosition
 
     public EntitySection(Class<T> entityClass, Visibility status) {
         this.chunkStatus = status;
         this.storage = new ClassInstanceMultiMap<>(entityClass);
     }
 
+    // Fusion start - Lithium: Util EntitySectionPosition
+    public void setPos(long chunkSectionPos) {
+        this.pos = chunkSectionPos;
+    }
+
+    public long getPos() {
+        return this.pos;
+    }
+    // Fusion end
+
     public void add(T entity) {
         this.storage.add(entity);
     }
diff --git a/src/main/java/net/minecraft/world/level/entity/EntitySectionStorage.java b/src/main/java/net/minecraft/world/level/entity/EntitySectionStorage.java
index 6026a75057d1e2f49d9ee4075a204021c092da77..1afa1ce43beb810634e7500a86023e7db0ee76ef 100644
--- a/src/main/java/net/minecraft/world/level/entity/EntitySectionStorage.java
+++ b/src/main/java/net/minecraft/world/level/entity/EntitySectionStorage.java
@@ -169,7 +169,11 @@ public class EntitySectionStorage<T extends EntityAccess> {
         long l = getChunkKeyFromSectionKey(sectionPos);
         Visibility visibility = this.intialSectionVisibility.get(l);
         this.sectionIds.add(sectionPos);
-        return new EntitySection<>(this.entityClass, visibility);
+        // Fusion start - Lithium: Util EntitySectionPosition
+        EntitySection section = new EntitySection<>(this.entityClass, visibility);
+        section.setPos(sectionPos);
+        return section;
+        // Fusion end
     }
 
     public LongSet getAllChunksWithExistingSections() {
