From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Tue, 4 Oct 2022 20:04:31 +0900
Subject: [PATCH] Fusion-Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a7e2fab3e174fe2c1d79d904f7eaff88dd0c3c82..0ca7d8cef78e4ad204240f3372537935de199b91 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1981,6 +1981,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Fusion start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getFusionConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Fusion end
+
         /**
          * Sends the component to the player
          *
