From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Tue, 4 Oct 2022 20:04:31 +0900
Subject: [PATCH] Fusion-Configuration


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e0f69edf603c2ec99bc92b16b18912272cc41bd9..ee6123a277d88e2368921a83ce697bd85340020e 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1983,6 +1983,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Fusion start
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getFusionConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Fusion end
+
         /**
          * Sends the component to the player
          *
