From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 17 Oct 2022 20:47:29 +0900
Subject: [PATCH] Fusion-LithiumConfiguration


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index d409506a4aa19faa3dc8cb36a48d1949a23a8f8f..ab87e88c425e792edb4fe1c07a1cf48b1d91abdf 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -1,5 +1,6 @@
 package com.github.ipecter.fusion;
 
+import com.github.ipecter.fusion.configuraions.LithiumConfig;
 import com.google.common.base.Throwables;
 import com.google.common.collect.ImmutableMap;
 import net.minecraft.server.MinecraftServer;
@@ -60,6 +61,7 @@ public class FusionConfig {
         set("config-version", 1);
 
         readConfig(FusionConfig.class, null);
+        readConfig(LithiumConfig.class, null);
     }
 
     protected static void log(String s) {
diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index 29715bcbad34d2cbdfab67d0db1739a408d564a8..3ac4805af970f355619a7699a702bce63de99d4c 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -1,5 +1,6 @@
 package com.github.ipecter.fusion;
 
+import com.github.ipecter.fusion.configuraions.LithiumWorldConfig;
 import org.apache.commons.lang.BooleanUtils;
 import org.simpleyaml.configuration.ConfigurationSection;
 import org.simpleyaml.configuration.comments.CommentType;
@@ -23,6 +24,7 @@ public class FusionWorldConfig {
     public void init() {
         log("-------- World Settings For [" + worldName + "] --------");
         FusionConfig.readConfig(FusionWorldConfig.class, this);
+        FusionConfig.readConfig(LithiumWorldConfig.class, this);
     }
 
     protected static void setComment(String path, String... comment){
diff --git a/src/main/java/com/github/ipecter/fusion/configuraions/LithiumConfig.java b/src/main/java/com/github/ipecter/fusion/configuraions/LithiumConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..c9ba665e2ade6c29168ada8da8b49ec6c64ecbcf
--- /dev/null
+++ b/src/main/java/com/github/ipecter/fusion/configuraions/LithiumConfig.java
@@ -0,0 +1,17 @@
+package com.github.ipecter.fusion.configuraions;
+
+import com.github.ipecter.fusion.FusionConfig;
+
+public class LithiumConfig extends FusionConfig {
+
+    public LithiumConfig(){
+        super();
+    }
+
+    public static boolean lithiumEnable = true;
+    private static void lithiumSettings() {
+        setComment("lithium", "[ Lithium ] General-Purpose Optimization Mod");
+        lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
+    }
+
+}
diff --git a/src/main/java/com/github/ipecter/fusion/configuraions/LithiumWorldConfig.java b/src/main/java/com/github/ipecter/fusion/configuraions/LithiumWorldConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..24bf6c22afabd2e8734cf9875e557b05b193795f
--- /dev/null
+++ b/src/main/java/com/github/ipecter/fusion/configuraions/LithiumWorldConfig.java
@@ -0,0 +1,14 @@
+package com.github.ipecter.fusion.configuraions;
+
+import com.github.ipecter.fusion.FusionWorldConfig;
+
+public class LithiumWorldConfig extends FusionWorldConfig {
+
+    private final String worldName;
+
+    public LithiumWorldConfig(String worldName) {
+        super(worldName);
+        this.worldName = worldName;
+    }
+
+}
