From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 5 Oct 2022 16:13:46 +0900
Subject: [PATCH] Fusion-LithiumConfiguration


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index d1999cf58c4a39f74464d8c055d7a8cd9de0d9b8..3f718ec4fdcafff8c0cf878184c2830dac202a50 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -194,4 +194,10 @@ public class FusionConfig {
         }
         return builder.build();
     }
+
+    public static boolean lithiumEnable = true;
+    private static void lithiumSettings() {
+        setComment("lithium", "[ Lithium ] General-Purpose Optimization Mod");
+        lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
+    }
 }
\ No newline at end of file
