From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 19 Oct 2022 18:48:32 +0900
Subject: [PATCH] a2


diff --git a/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java b/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java
index cb01bddee1a5138a1cc79a393f43bca73e0491f9..335f37ba7283bdfdd2229a0d1ec30545924c6525 100644
--- a/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java
@@ -8,9 +8,11 @@ public class LithiumConfig extends FusionConfig {
         setComment("lithium", "[ Lithium ] General-Purpose Optimization Mod");
         enable = getBoolean("lithium.enable", true);
         fastUtil = getBoolean("lithium.fast-util", true);
+        hashedReferenceList = getBoolean("lithium.hashed-reference-list", true);
     }
 
     public boolean enable;
     public boolean fastUtil;
+    public boolean hashedReferenceList;
 
 }
