From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 19 Oct 2022 18:49:41 +0900
Subject: [PATCH] a4


diff --git a/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java b/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java
index 512e33b9a889347c1e52cf30a6b234cc86d4c3db..3cb59bcf4b4924120710b5b1f4562412f00cd918 100644
--- a/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/configurations/LithiumConfig.java
@@ -10,11 +10,13 @@ public class LithiumConfig extends FusionConfig {
         fastUtil = getBoolean("lithium.fast-util", true);
         hashedReferenceList = getBoolean("lithium.hashed-reference-list", true);
         compactSineLUT = getBoolean("lithium.compact-sine-lut", true);
+        fastRetrieval = getBoolean("lithium.fast-retrieval", true);
     }
 
     public boolean enable;
     public boolean fastUtil;
     public boolean hashedReferenceList;
     public boolean compactSineLUT;
+    public boolean fastRetrieval;
 
 }
