From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Tue, 1 Nov 2022 09:46:28 +0900
Subject: [PATCH] Lithium-CollectionsAttributes


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index e593b6c9bee23da1e1040ea12097bae2c750f78d..7122ecf8673c0196fce682d8907dae75e6826b21 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -206,6 +206,7 @@ public class FusionConfig {
     public static boolean lithiumEntityFastElytraCheck = true;
     public static boolean lithiumEntityFastHandSwing = true;
     public static boolean lithiumEntityFastPowderSnowCheck = true;
+    public static boolean lithiumCollectionsAttributes = true;
 
     private static void lithiumConfig() {
         setComment("lithium", "[ Lithium ] General-Purpose Optimization Mod");
@@ -220,5 +221,6 @@ public class FusionConfig {
         lithiumEntityFastElytraCheck = getBoolean("lithium.entity.fast-elytra-check", lithiumEntityFastElytraCheck) && lithiumEnable;
         lithiumEntityFastHandSwing = getBoolean("lithium.entity.fast-hand-swing", lithiumEntityFastHandSwing) && lithiumEnable;
         lithiumEntityFastPowderSnowCheck = getBoolean("lithium.entity.fast-powder-snow-check", lithiumEntityFastPowderSnowCheck) && lithiumEnable;
+        lithiumCollectionsAttributes = getBoolean("lithium.collections.attributes", lithiumCollectionsAttributes) && lithiumEnable;
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
index 00fc98797aea23e1f586b8e7f85fc27e2019352f..a0372057239d4e93a485adc3dd6d2f68133efccf 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -19,8 +19,8 @@ import org.slf4j.Logger;
 
 public class AttributeMap {
     private static final Logger LOGGER = LogUtils.getLogger();
-    private final Map<Attribute, AttributeInstance> attributes = Maps.newHashMap();
-    private final Set<AttributeInstance> dirtyAttributes = Sets.newHashSet();
+    private final Map<Attribute, AttributeInstance> attributes = com.github.ipecter.fusion.FusionConfig.lithiumCollectionsAttributes ? new it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap<>(0) : Maps.newHashMap(); // Fusion - Lithium: Collections Attributes
+    private final Set<AttributeInstance> dirtyAttributes = com.github.ipecter.fusion.FusionConfig.lithiumCollectionsAttributes ? new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(0) : Sets.newHashSet(); // Fusion - Lithium: Collections Attributes
     private final AttributeSupplier supplier;
     private final net.minecraft.world.entity.LivingEntity entity; // Purpur
     private final java.util.function.Function<Attribute, AttributeInstance> createInstance; // Pufferfish
