From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Thu, 27 Oct 2022 14:44:59 +0900
Subject: [PATCH] Lithium-Entity-FastHandSwing

Mirai - etil2jz - GPL 3.0
Lithium - CaffeineMC  - GPL 3.0

diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index bc0d8d9695c3fdac59b040fd05ac90ba4ccd7070..74e676a0e16d455d50eb5c786170c1a381e9f7b6 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -133,11 +133,13 @@ public class FusionWorldConfig {
     public boolean lithiumEnable = true;
     public boolean lithiumCollectionsTileEntityTickers = true;
     public boolean lithiumEntityFastElytraCheck = true;
+    public boolean lithiumEntityFastHandSwing = true;
 
     private void lithiumConfig() {
         lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
         lithiumCollectionsTileEntityTickers = getBoolean("lithium.collections.tile-entity-tickers", lithiumCollectionsTileEntityTickers) && lithiumEnable;
         lithiumEntityFastElytraCheck = getBoolean("lithium.entity.fast-elytra-check", lithiumEntityFastElytraCheck) && lithiumEnable;
+        lithiumEntityFastHandSwing = getBoolean("lithium.entity.fast-hand-swing", lithiumEntityFastHandSwing) && lithiumEnable;
     }
 
     private void lithiumConfigSetup() {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 0c9aca5186b1fdb7e7d3dcb8ee41223551383572..4f9746675160cd9986514b93f4dc25171a36024b 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2549,6 +2549,8 @@ public abstract class LivingEntity extends Entity {
     }
 
     protected void updateSwingTime() {
+		if (this.level.fusionConfig.lithiumEntityFastHandSwing && !this.swinging && this.swingTime == 0) return; // Fusion - Lithium: Entity Fast Hand Swing
+
         int i = this.getCurrentSwingDuration();
 
         if (this.swinging) {
