From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 28 Dec 2022 15:05:35 +0900
Subject: [PATCH] Lithium-Entity-FastElytraCheck


diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index 7c14500f529c9d9869a74f2dbe0a01c0594ecb3e..b7b030d3679434f02de31ec91bbeca83e5e5e332 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -132,10 +132,12 @@ public class FusionWorldConfig {
 
     public boolean lithiumEnable = true;
     public boolean lithiumCollectionsBlockEntityTickers = true;
+    public boolean lithiumEntityFastElytraCheck = true;
 
     private void lithiumConfig() {
         lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
         lithiumCollectionsBlockEntityTickers = getBoolean("lithium.collections.block-entity-tickers", lithiumCollectionsBlockEntityTickers) && lithiumEnable;
+        lithiumEntityFastElytraCheck = getBoolean("lithium.entity.fast-elytra-check", lithiumEntityFastElytraCheck) && lithiumEnable;
     }
 
     private void lithiumConfigSetup() {
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 094afd5636590608c8ec31c9fec97614dfef60db..c0b5e74b79cfb0fd5028235cd2f7a59b3c7f0747 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3581,6 +3581,8 @@ public abstract class LivingEntity extends Entity {
     }
 
     private void updateFallFlying() {
+        if (this.level.fusionConfig.lithiumEntityFastElytraCheck && !this.isFallFlying()) return; // Fusion - Lithium: Entity FastElytraCheck
+
         boolean flag = this.getSharedFlag(7);
 
         if (flag && !this.onGround && !this.isPassenger() && !this.hasEffect(MobEffects.LEVITATION)) {
