From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 28 Dec 2022 15:23:53 +0900
Subject: [PATCH] Lithium-Collections-EntityAttributes


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index b36f3b07b0f855b2c2ec8a0fc8fea69c6e93fe42..c716273b1b6f27a45a6557505895fbefe25b4a53 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -225,6 +225,7 @@ public class FusionConfig {
     public static boolean lithiumCollectionsEntityAIGoals = true;
     public static boolean lithiumCollectionsStoreGameRules = true;
     public static boolean lithiumShapesPrecomputeShapeArrays = true;
+    public static boolean lithiumCollectionsEntityAttributes = true;
 
     private static void lithiumConfig() {
         setComment("lithium", "[ Lithium ] General-Purpose Optimization Mod");
@@ -235,6 +236,7 @@ public class FusionConfig {
         lithiumCollectionsEntityAIGoals = getBoolean("lithium.collections.entity-ai-goals", lithiumCollectionsEntityAIGoals) && lithiumEnable;
         lithiumCollectionsStoreGameRules = getBoolean("lithium.collections.store-gamerules", lithiumCollectionsStoreGameRules) && lithiumEnable;
         lithiumShapesPrecomputeShapeArrays = getBoolean("lithium.shapes.precompute-shape-arrays", lithiumShapesPrecomputeShapeArrays) && lithiumEnable;
+        lithiumCollectionsEntityAttributes = getBoolean("lithium.collections.entity-attributes", lithiumCollectionsEntityAttributes) && lithiumEnable;
     }
 
     private static void lithiumConfigSetup() {
diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index 24140bd78be47d987ffaa55f9f118f861d5b996d..7e85d00c9ce2c8ebd6cca9d6c8e7fcaccea6c17b 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -135,6 +135,7 @@ public class FusionWorldConfig {
     public boolean lithiumEntityFastElytraCheck = true;
     public boolean lithiumEntityFastHandSwing = true;
     public boolean lithiumEntityFastPowderSnowCheck = true;
+    public boolean lithiumCollectionsEntityAttributes = true;
 
     private void lithiumConfig() {
         lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
@@ -142,6 +143,7 @@ public class FusionWorldConfig {
         lithiumEntityFastElytraCheck = getBoolean("lithium.entity.fast-elytra-check", lithiumEntityFastElytraCheck) && lithiumEnable;
         lithiumEntityFastHandSwing = getBoolean("lithium.entity.fast-hand-swing", lithiumEntityFastHandSwing) && lithiumEnable;
         lithiumEntityFastPowderSnowCheck = getBoolean("lithium.entity.fast-powder-snow-check", lithiumEntityFastPowderSnowCheck) && lithiumEnable;
+        lithiumCollectionsEntityAttributes = getBoolean("lithium.collections.entity-attributes", lithiumCollectionsEntityAttributes) && lithiumEnable;
     }
 
     private void lithiumConfigSetup() {
diff --git a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
index 210a0bee1227e4671909dd553ab22027cfc868fb..6b249278565a577308618483657584262ee96429 100644
--- a/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
+++ b/src/main/java/net/minecraft/world/entity/ai/attributes/AttributeMap.java
@@ -20,8 +20,8 @@ import org.slf4j.Logger;
 
 public class AttributeMap {
     private static final Logger LOGGER = LogUtils.getLogger();
-    private final Map<Attribute, AttributeInstance> attributes = Maps.newHashMap();
-    private final Set<AttributeInstance> dirtyAttributes = Sets.newHashSet();
+    private final Map<Attribute, AttributeInstance> attributes; // Fusion - Lithium: Collections EntityAttributes
+    private final Set<AttributeInstance> dirtyAttributes; // Fusion - Lithium: Collections EntityAttributes
     private final AttributeSupplier supplier;
     private final java.util.function.Function<Attribute, AttributeInstance> createInstance; // Pufferfish
     private final net.minecraft.world.entity.LivingEntity entity; // Purpur
@@ -33,6 +33,15 @@ public class AttributeMap {
     public AttributeMap(AttributeSupplier defaultAttributes, net.minecraft.world.entity.LivingEntity entity) {
         this.entity = entity;
         // Purpur end
+        // Fusion start - Lithium: Collections EntityAttributes
+        if (this.entity != null){
+            this.attributes = this.entity.level.fusionConfig.lithiumCollectionsEntityAttributes ? new it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap<>(0) : Maps.newHashMap();
+            this.dirtyAttributes = this.entity.level.fusionConfig.lithiumCollectionsEntityAttributes ? new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(0) : Sets.newHashSet();
+        } else {
+            this.attributes = com.github.ipecter.fusion.FusionConfig.lithiumCollectionsEntityAttributes ? new it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap<>(0) : Maps.newHashMap();
+            this.dirtyAttributes = com.github.ipecter.fusion.FusionConfig.lithiumCollectionsEntityAttributes ? new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(0) : Sets.newHashSet();
+        }
+        // Fusion end
         this.supplier = defaultAttributes;
         this.createInstance = attribute -> this.supplier.createInstance(this::onAttributeModified, attribute); // Pufferfish
     }
