From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 9 Nov 2022 19:58:50 +0900
Subject: [PATCH] Lithium-Entity-AISecondaryPoiSensor


diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index a9f46db1bc7031aef515e131f22b8e252d9d32c4..e25c7c4b991a21b26036234eb0422047cb24845f 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -131,6 +131,7 @@ public class FusionWorldConfig {
     public boolean lithiumEntityRaiderAI = true;
     public boolean lithiumBlockMovingBlockEntityShapes = true;
     public boolean lithiumChunkFastAccess = true;
+    public boolean lithiumEntityAISecondaryPoiSensor = true;
 
     private void lithiumConfig() {
         lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
@@ -142,5 +143,6 @@ public class FusionWorldConfig {
         lithiumEntityRaiderAI = getBoolean("lithium.entity.raider-ai", lithiumEntityRaiderAI) && lithiumEnable;
         lithiumBlockMovingBlockEntityShapes = getBoolean("lithium.block.moving-block-entity-shapes", lithiumBlockMovingBlockEntityShapes) && lithiumEnable;
         lithiumChunkFastAccess = getBoolean("lithium.chunk.fast-access", lithiumChunkFastAccess) && lithiumEnable;
+        lithiumEntityAISecondaryPoiSensor = getBoolean("lithium.entity.ai-secondary-poi-sensor", lithiumEntityAISecondaryPoiSensor) && lithiumEnable;
     }
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java b/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
index d2703432af207c74ea8d298a784329c3219d2f13..8999e183d95d33493fe3b0a60f9f014ba0683004 100644
--- a/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -11,6 +11,7 @@ import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.ai.Brain;
 import net.minecraft.world.entity.ai.memory.MemoryModuleType;
 import net.minecraft.world.entity.npc.Villager;
+import net.minecraft.world.entity.npc.VillagerProfession;
 import net.minecraft.world.level.Level;
 
 public class SecondaryPoiSensor extends Sensor<Villager> {
@@ -24,7 +25,8 @@ public class SecondaryPoiSensor extends Sensor<Villager> {
     protected void doTick(ServerLevel world, Villager entity) {
         // Purpur start - make sure clerics don't wander to soul sand when the option is off
         Brain<?> brain = entity.getBrain();
-        if (!world.purpurConfig.villagerClericsFarmWarts && entity.getVillagerData().getProfession() == net.minecraft.world.entity.npc.VillagerProfession.CLERIC) {
+        VillagerProfession profession = entity.getVillagerData().getProfession(); // Fusion - Short Code
+        if ((!world.purpurConfig.villagerClericsFarmWarts && profession == net.minecraft.world.entity.npc.VillagerProfession.CLERIC) || (entity.level.fusionConfig.lithiumEntityAISecondaryPoiSensor && profession.secondaryPoi().isEmpty())) { // Fusion - Lithium: Entity Secondary Poi Sensor
             brain.eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
             return;
         }
