From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Thu, 5 Jan 2023 15:11:03 +0900
Subject: [PATCH] Lithium-AI-Sensor-SecondaryPoi

Lithium - CaffeineMC  - GPL 3.0
Mirai - etil2jz - GPL 3.0

diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index e7371b42872e69e05bfb8fca14d1f2a9302d7e04..111d5e350ec5b13b7ade1c451c518d0814f5d67e 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -138,6 +138,7 @@ public class FusionWorldConfig {
     public boolean lithiumAIRaid = true;
     public boolean lithiumBlockMovingBlockShapes = true;
     public boolean lithiumGenChunkRegion = true;
+    public boolean lithiumAISensorSecondaryPoi = true;
 
     private void lithiumConfig() {
         lithiumCollectionsBlockEntityTickers = getBoolean("lithium.collections.block-entity-tickers", lithiumCollectionsBlockEntityTickers) && lithiumEnable;
@@ -147,6 +148,7 @@ public class FusionWorldConfig {
         lithiumAIRaid = getBoolean("lithium.ai.raid", lithiumAIRaid) && lithiumEnable;
         lithiumBlockMovingBlockShapes = getBoolean("lithium.block.moving-block-shapes", lithiumBlockMovingBlockShapes) && lithiumEnable;
         lithiumGenChunkRegion = getBoolean("lithium.gen.chunk-region", lithiumGenChunkRegion) && lithiumEnable;
+        lithiumAISensorSecondaryPoi = getBoolean("lithium.ai.sensor.secondary-poi", lithiumAISensorSecondaryPoi) && lithiumEnable;
     }
 
     private void lithiumConfigSetup() {
diff --git a/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java b/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
index d2703432af207c74ea8d298a784329c3219d2f13..019487e037caff2c1d5b736c014825b55842f3d7 100644
--- a/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/src/main/java/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -24,7 +24,7 @@ public class SecondaryPoiSensor extends Sensor<Villager> {
     protected void doTick(ServerLevel world, Villager entity) {
         // Purpur start - make sure clerics don't wander to soul sand when the option is off
         Brain<?> brain = entity.getBrain();
-        if (!world.purpurConfig.villagerClericsFarmWarts && entity.getVillagerData().getProfession() == net.minecraft.world.entity.npc.VillagerProfession.CLERIC) {
+        if ((world.fusionConfig.lithiumAISensorSecondaryPoi && entity.getVillagerData().getProfession().secondaryPoi().isEmpty()) || (!world.purpurConfig.villagerClericsFarmWarts && entity.getVillagerData().getProfession() == net.minecraft.world.entity.npc.VillagerProfession.CLERIC)) { // Fusion - Lithium: AI Sensor SecondaryPoi
             brain.eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
             return;
         }
