From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 9 Jan 2023 19:31:19 +0900
Subject: [PATCH] Fusion-C2MEConfiguration


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index 6529508d8b7c6559382af763f2f78e283e3d55ec..0c77915177533f30bd95b53dcac03937dd162578 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -249,4 +249,14 @@ public class FusionConfig {
     private static void lithiumConfigSetup() {
         lithiumWorldTickScheduler = getBoolean("lithium.world.tick-scheduler", lithiumWorldTickScheduler) && lithiumEnable;
     }
+
+    public static boolean c2meEnable = true;
+
+    private static void c2meConfig() {
+        setComment("c2me", "[ C2ME ] Chunk Performance Improve Mod");
+        c2meEnable = getBoolean("c2me.enable", c2meEnable);
+    }
+
+    private static void c2meConfigSetup() {
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index e2cec470613d5f8558a281889859d9f03d403baf..4425bf963d7d3fc20ad32f369bb04e31d2d1aad1 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -158,4 +158,13 @@ public class FusionWorldConfig {
 
     private void lithiumConfigSetup() {
     }
+
+    public boolean c2meEnable = true;
+
+    private void c2meConfig() {
+        c2meEnable = getBoolean("c2me.enable", c2meEnable);
+    }
+
+    private void c2meConfigSetup() {
+    }
 }
\ No newline at end of file
