From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 14 Nov 2022 15:34:23 +0900
Subject: [PATCH] Fusion-C2MEConfiguration


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index 0af30ef2460ddd34cade886b9584a73e120e8c14..143db84d4e870c97e2b032fe384eff631bfb5de1 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -239,4 +239,14 @@ public class FusionConfig {
     private static void lithiumConfigSetup() {
         lithiumWorldImproveWorldTickScheduler = getBoolean("lithium.world.improve-tick-scheduler", lithiumWorldImproveWorldTickScheduler) && lithiumEnable;
     }
+
+    public static boolean c2meEnable = true;
+
+    private static void c2meConfig() {
+        setComment("c2me", "[ C2ME ] Chunk Performance Improve Mod");
+        c2meEnable = getBoolean("c2me.enable", c2meEnable);
+    }
+
+    private static void c2meConfigSetup() {
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index 9a9f9daad5e552eb1310e6039f109fe50e43535a..8ed530e5fba8c421324fe7321ddcf43b9b7b93d1 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -158,4 +158,13 @@ public class FusionWorldConfig {
 
     private void lithiumConfigSetup() {
     }
+
+    public boolean c2meEnable = true;
+
+    private void c2meConfig() {
+        c2meEnable = getBoolean("c2me.enable", c2meEnable);
+    }
+
+    private void c2meConfigSetup() {
+    }
 }
\ No newline at end of file
