From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 9 Jan 2023 12:25:36 +0900
Subject: [PATCH] Lithium-PositionedEntityTrackingSection


diff --git a/src/main/java/me/jellysquid/mods/lithium/common/entity/PositionedEntityTrackingSection.java b/src/main/java/me/jellysquid/mods/lithium/common/entity/PositionedEntityTrackingSection.java
new file mode 100644
index 0000000000000000000000000000000000000000..b5fc6ad5c7c4efe3bbd3ec71ae1e3ef3bf3c578e
--- /dev/null
+++ b/src/main/java/me/jellysquid/mods/lithium/common/entity/PositionedEntityTrackingSection.java
@@ -0,0 +1,7 @@
+package me.jellysquid.mods.lithium.common.entity;
+
+public interface PositionedEntityTrackingSection {
+    void setPos(long chunkSectionPos);
+
+    long getPos();
+}
\ No newline at end of file
