From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Tue, 10 Jan 2023 10:12:06 +0900
Subject: [PATCH] C2ME-PooledFeatureContext


diff --git a/src/main/java/com/ishland/c2me/opts/allocs/common/PooledFeatureContext.java b/src/main/java/com/ishland/c2me/opts/allocs/common/PooledFeatureContext.java
new file mode 100644
index 0000000000000000000000000000000000000000..3139e2d264e45258ed5ae546b6dd3b3ff46880be
--- /dev/null
+++ b/src/main/java/com/ishland/c2me/opts/allocs/common/PooledFeatureContext.java
@@ -0,0 +1,69 @@
+package com.ishland.c2me.opts.allocs.common;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.level.WorldGenLevel;
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
+import net.minecraft.world.level.levelgen.feature.configurations.FeatureConfiguration;
+import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext;
+
+import java.util.Optional;
+
+public class PooledFeatureContext<FC extends FeatureConfiguration> extends FeaturePlaceContext<FC> {
+
+    public static final ThreadLocal<SimpleObjectPool<PooledFeatureContext<?>>> POOL = ThreadLocal.withInitial(() -> new SimpleObjectPool<>(unused -> new PooledFeatureContext<>(), unused -> {}, 2048));
+
+    private Optional<ConfiguredFeature<?, ?>> feature;
+    private WorldGenLevel world;
+    private ChunkGenerator generator;
+    private RandomSource random;
+    private BlockPos origin;
+    private FC config;
+
+    public PooledFeatureContext() {
+        super(null, null, null, null, null, null);
+    }
+
+    public void reInit(Optional<ConfiguredFeature<?, ?>> feature, WorldGenLevel world, ChunkGenerator generator, RandomSource random, BlockPos origin, FC config) {
+        this.feature = feature;
+        this.world = world;
+        this.generator = generator;
+        this.random = random;
+        this.origin = origin;
+        this.config = config;
+    }
+
+    public void reInit() {
+        this.feature = null;
+        this.world = null;
+        this.generator = null;
+        this.random = null;
+        this.origin = null;
+        this.config = null;
+    }
+
+    public WorldGenLevel getWorld() {
+        return this.world;
+    }
+
+    public ChunkGenerator getGenerator() {
+        return this.generator;
+    }
+
+    public RandomSource getRandom() {
+        return this.random;
+    }
+
+    public BlockPos getOrigin() {
+        return this.origin;
+    }
+
+    public FC getConfig() {
+        return this.config;
+    }
+
+    public Optional<ConfiguredFeature<?, ?>> getFeature() {
+        return this.feature;
+    }
+}
\ No newline at end of file
