From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 14 Nov 2022 16:46:20 +0900
Subject: [PATCH] Fusion-PacketSettingsConfiguration


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index 9a7cf0d4540d7de4060a60922aa1e7624705f68a..1db413994e8d7186f0ee9631966fb1572c9bce88 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -251,4 +251,10 @@ public class FusionConfig {
 
     private static void c2meConfigSetup() {
     }
+
+    private static void packetSettings() {
+    }
+
+    private static void packetSettingsSetup() {
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index 8efbdfc5ef48b329988bc90606c7e0fccfd504fd..df2695564698ac75e433348c8b6ed6deb3ca01a4 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -169,4 +169,10 @@ public class FusionWorldConfig {

     private void c2meConfigSetup() {
     }
+
+    private void packetSettings() {
+    }
+
+    private void packetSettingsSetup() {
+    }
 }
\ No newline at end of file
