From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 14 Nov 2022 16:46:20 +0900
Subject: [PATCH] Fusion-PacketSettingsConfiguration


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index 9a7cf0d4540d7de4060a60922aa1e7624705f68a..480c6d53d769b36f29801a913f489d75594addd4 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -251,4 +251,11 @@ public class FusionConfig {
 
     private static void c2meConfigSetup() {
     }
+
+
+    private static void packetSettings() {
+    }
+
+    private static void packetSettingsSetup() {
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index 8efbdfc5ef48b329988bc90606c7e0fccfd504fd..566dadfe1f0461ff7e8e46a2467f364a24a5147d 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -169,4 +169,11 @@ public class FusionWorldConfig {
 
     private void c2meConfigSetup() {
     }
+
+
+    private void packetSettings() {
+    }
+
+    private void packetSettingsSetup() {
+    }
 }
\ No newline at end of file
