From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Sun, 19 Feb 2023 18:53:43 +0900
Subject: [PATCH] 1


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index db64047fd306a66d14c134bf230efb2272be18b7..53eb9ec4460dbccc9d8e810bc2b5246887060a0d 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -257,4 +257,14 @@ public class FusionConfig {
         setComment("lithium", "[ Lithium ] General-Purpose Optimization Mod");
         lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
     }
+
+    public static boolean c2meEnable = true;
+
+    private static void c2meConfig() {
+    }
+
+    private static void c2meConfigSetup() {
+        setComment("c2me", "[ C2ME ] Chunk Performance Improve Mod");
+        c2meEnable = getBoolean("c2me.enable", c2meEnable);
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
index de9b22a20d1c85e10da6844a461b7fa99c3b6957..140aa7c9ee141c914f084cad1fab6172f13173b4 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionWorldConfig.java
@@ -158,4 +158,13 @@ public class FusionWorldConfig {
     private void lithiumConfigSetup() {
         lithiumEnable = getBoolean("lithium.enable", lithiumEnable);
     }
+
+    public boolean c2meEnable = true;
+
+    private void c2meConfig() {
+    }
+
+    private void c2meConfigSetup() {
+        c2meEnable = getBoolean("c2me.enable", c2meEnable);
+    }
 }
\ No newline at end of file
