From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 14 Nov 2022 17:24:00 +0900
Subject: [PATCH] Suki-SendMorePacketsImmediately

Suki - SuCraft - GPL 3.0

diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index b11d33296d6f29a74302360bb353c3e6a6e1f658..e72aba64b15c870a7d6fd78a9ab090bd295cc38e 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -253,9 +253,11 @@ public class FusionConfig {
     }

     public static boolean packetFlushAfterMorePackets = true;
+    public static boolean packetSendMorePacketsImmediately = true;

     private static void packetSettings() {
         packetFlushAfterMorePackets = getBoolean("packet.flush-after-more-packets", packetFlushAfterMorePackets);
+        packetSendMorePacketsImmediately = getBoolean("packet.send-more-packets-immediately", packetSendMorePacketsImmediately);
     }

     private static void packetSettingsSetup() {
diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index ded646f98f95556ac1253bdfbc441c32b8509d76..bc5b77a475223cbd946e34f1b99d84badb184f0b 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -356,7 +356,12 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
                 packet instanceof net.minecraft.network.protocol.game.ClientboundSetActionBarTextPacket ||
                 packet instanceof net.minecraft.network.protocol.game.ClientboundSetTitlesAnimationPacket ||
                 packet instanceof net.minecraft.network.protocol.game.ClientboundClearTitlesPacket ||
-                packet instanceof net.minecraft.network.protocol.game.ClientboundBossEventPacket;
+                packet instanceof net.minecraft.network.protocol.game.ClientboundBossEventPacket ||
+                // Fusion start - Suki: Send More Packets Immediately
+                    ( com.github.ipecter.fusion.FusionConfig.packetSendMorePacketsImmediately && (packet instanceof net.minecraft.network.protocol.game.ClientboundBossEventPacket ||
+                packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerInfoPacket ||
+                packet instanceof net.minecraft.network.protocol.game.ClientboundMapItemDataPacket));
+                // Fusion end
         }
         // Paper end
     }
