From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 2 Jan 2023 02:48:35 +0900
Subject: [PATCH] Fusion-NoChatReportsConfiguration


diff --git a/src/main/java/com/github/ipecter/fusion/FusionConfig.java b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
index 62a83e77c6539b1e5bd4c9ade3350326b9a8398d..7d09004e666aaabc70edc04c4f9f442601fc9207 100644
--- a/src/main/java/com/github/ipecter/fusion/FusionConfig.java
+++ b/src/main/java/com/github/ipecter/fusion/FusionConfig.java
@@ -203,4 +203,14 @@ public class FusionConfig {
         }
         return builder.build();
     }
+
+    public static boolean noChatReportsEnable = true;
+
+    private static void noChatReportsConfig() {
+        setComment("no-chat-reports", "[ No Chat Reports ] Player Chat Reporting Control Mod");
+        noChatReportsEnable = getBoolean("no-chat-reports.enable", noChatReportsEnable);
+    }
+
+    private static void noChatReportsConfigSetup() {
+    }
 }
\ No newline at end of file
